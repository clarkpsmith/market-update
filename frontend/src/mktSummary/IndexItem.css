.IndexItem-title {
  margin-bottom: 3px;
}
