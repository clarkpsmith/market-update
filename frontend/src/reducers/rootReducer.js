import {
  REMOVE_TICKER,
  ADD_TICKER,
  UPDATE_CURR_USER,
  LOG_OUT,
  SIGN_UP,
  UPDATE_PROFILE,
  REFRESH_TICKER,
  SET_WATCHLIST_DATA,
  SEARCH_TICKER_DATA,
} from "../actions/types";

const DEFAULT_STATE = {
  currentUser: {},
  watchlistData: [],
  searchTickerData: null,
};

function rootReducer(state = DEFAULT_STATE, action) {
  switch (action.type) {
    case REMOVE_TICKER:
      return {
        ...state,
        currentUser: {
          ...state.currentUser,
          watchlist: state.currentUser.watchlist.filter(
            (t) => t !== action.ticker.symbol
          ),
        },
        watchlistData: state.watchlistData.filter(
          (t) => t.symbol !== action.ticker.symbol
        ),
      };
    case SEARCH_TICKER_DATA:
      return {
        ...state,
        searchTickerData: action.tickerData,
      };

    case ADD_TICKER:
      return {
        ...state,
        currentUser: {
          ...state.currentUser,
          watchlist: [...state.currentUser.watchlist, action.ticker.symbol],
        },
        watchlistData: [...state.watchlistData, action.ticker],
      };

    case UPDATE_CURR_USER:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case UPDATE_PROFILE:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case LOG_OUT:
      return {
        ...state,
        currentUser: {},
        watchlistData: [],
      };

    case SIGN_UP:
      return {
        ...state,
        currentUser: { ...action.currentUser },
      };

    case SET_WATCHLIST_DATA:
      return {
        ...state,
        watchlistData: [...action.watchlistData],
      };

    case REFRESH_TICKER:
      return {
        ...state,
        watchlistData: [
          ...state.watchlistData.map((t) => {
            if (t.symbol === action.ticker.symbol) return action.ticker;
            else return t;
          }),
        ],
      };
    default:
      return state;
  }
}
export default rootReducer;
