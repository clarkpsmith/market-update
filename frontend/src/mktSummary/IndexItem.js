import React, { useState } from "react";
import { addCommas } from "../helpers/abreviateLargeNums";
import Indexinfo from "./IndexInfo";
import { useDispatch } from "react-redux";
import { refreshTicker } from "../actions/actionCreators";
import "../stockinfo/Stockinfo.css";
import "../watchlist/Item.css";
import "./IndexItem.css";

const IndexItem = ({ ticker }) => {
  const dispatch = useDispatch();
  const [expand, setExpand] = useState(false);
  const [marketChange, setMarketChange] = useState({
    percent: `${ticker.regularMarketChangePercent.toFixed(2)} %`,
    market: ticker.regularMarketChange.toFixed(2),
  });
  const [percent, setPercent] = useState(true);

  function handleClick() {
    percent ? setPercent(false) : setPercent(true);
  }

  function refresh() {
    dispatch(refreshTicker(ticker.symbol));
  }

  function handleExpand() {
    if (expand) setExpand(false);
    else setExpand(true);
  }

  let percentColor;
  marketChange.market >= 0
    ? (percentColor = "LimeGreen")
    : (percentColor = "red");
  const marketPrice = ticker.regularMarketPrice
    ? addCommas(ticker.regularMarketPrice.toFixed(2))
    : "N/A";

  if (!expand) {
    return (
      <div className="Item">
        <h4 className="Item-title IndexItem-title">
          <span
            className="Item-title-clickable"
            title="Expand"
            onClick={handleExpand}
          >
            {ticker.shortName}
          </span>
          <span>
            <span
              title="Current stock price, click to refresh"
              onClick={refresh}
              className="Stockinfo-price"
            >
              {marketPrice}
            </span>
          </span>
        </h4>
        <button
          onClick={handleClick}
          className="Stockinfo-daily-change Item-btn-change"
          style={{ backgroundColor: percentColor }}
        >
          {percent ? marketChange.percent : marketChange.market}
        </button>
      </div>
    );
  }
  if (expand) {
    return <Indexinfo ticker={ticker} />;
  }
};

export default IndexItem;
