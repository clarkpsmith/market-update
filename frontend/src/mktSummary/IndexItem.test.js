import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import IndexItem from "./IndexItem";
import { composeWithDevTools } from "redux-devtools-extension";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import rootReducer from "../reducers/rootReducer";
const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

const ticker = {
  averageDailyVolume3Month: 102300857,
  beta: 1.203743,
  dividendYield: 0.68,
  epsCurrentYear: 5.2,
  epsForward: 5.34,
  epsTrailingTwelveMonths: 4.449,
  fiftyTwoWeekHigh: 145.09,
  fiftyTwoWeekLow: 75.0525,
  forwardPE: 24.383896,
  fullExchangeName: "NasdaqGS",
  heldPercentInsiders: 0.076,
  longName: "Apple Inc.",
  marketCap: 2172892479488,
  pegRatio: 1.45,
  priceToSales: 6.677481,
  region: "US",
  regularMarketChange: 0.69000244,
  regularMarketDayHigh: 131.2582,
  regularMarketChangePercent: 0.5327381,
  regularMarketDayLow: 129.475,
  regularMarketOpen: 130.85,
  regularMarketPrice: 130.21,
  regularMarketVolume: 78973273,
  shortName: "Apple Inc.",
  symbol: "AAPL",
  trailingPE: 29.267254,
};

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <IndexItem ticker={ticker} />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <IndexItem ticker={ticker} />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
