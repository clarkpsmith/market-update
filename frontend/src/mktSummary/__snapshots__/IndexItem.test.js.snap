// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Item"
  >
    <h4
      class="Item-title IndexItem-title"
    >
      <span
        class="Item-title-clickable"
        title="Expand"
      >
        Apple Inc.
      </span>
      <span>
        <span
          class="Stockinfo-price"
          title="Current stock price, click to refresh"
        >
          130.21
        </span>
      </span>
    </h4>
    <button
      class="Stockinfo-daily-change Item-btn-change"
      style="background-color: LimeGreen;"
    >
      0.53 %
    </button>
  </div>
</DocumentFragment>
`;
